source/common/common/radix_tree.h

```diff
@@ -0,0 +1,324 @@
+
+      // Replace the original child with the split node
+      children_[firstChar] = std::move(splitNode);
+    }
+
+    /**
+     * Recursive helper for find operation.
+     * @param search the remaining search key.
+     * @param result the value to return if found.
+     * @return true if the key was found, false otherwise.
+     */
+    bool find_recursive(absl::string_view search, Value& result) const {
+      if (search.empty()) {
+        if (hasValue(*this)) {
+          result = value_;
+          return true;
+        }
+        return false;
+      }
+
+      uint8_t firstChar = static_cast<uint8_t>(search[0]);
+      auto childIt = children_.find(firstChar);
+      if (childIt == children_.end()) {
+        return false;
+      }
+
+      const RadixTreeNode& child = childIt->second;
+
+      // Check if the child's prefix matches the search
+      if (search.size() >= child.prefix_.size() &&
+          search.substr(0, child.prefix_.size()) == child.prefix_) {
+        absl::string_view newSearch(search.begin() + child.prefix_.size(), search.end());
+        return child.find_recursive(newSearch, result);
+      }
+
+      return false;
+    }
+  };
+
+  /**
+   * Get the child node for the given character key.
+   * @param node the node to get the child from.
+   * @param char_key the one-byte key of the child to get.
+   * @return pointer to the child node, or nullptr if not found.
+   */
+  const RadixTreeNode* getChild(const RadixTreeNode& node, uint8_t char_key) const {
+    auto it = node.children_.find(char_key);
+    if (it != node.children_.end()) {
+      return &it->second;
+    }
+    return nullptr;
+  }
+
+  /**
+   * Check if a node has a value (is a leaf node)
+   */
+  static bool hasValue(const RadixTreeNode& node) {
+    // For pointer types, check if the pointer is not null
+    if constexpr (std::is_pointer_v<Value>) {
+      return node.value_ != nullptr;
+    } else {
+      return static_cast<bool>(node.value_);
+    }
+  }
+
+  /**
+   * Find the longest common prefix between two strings
+   */
+  static size_t longestPrefix(absl::string_view a, absl::string_view b) {
+    size_t len = std::min(a.size(), b.size());
+    for (size_t i = 0; i < len; i++) {
+      if (a[i] != b[i]) {
+        return i;
+      }
+    }
+    return len;
+  }
+
+public:
+  /**
+   * Adds an entry to the RadixTree at the given Key.
+   * @param key the key used to add the entry.
+   * @param value the value to be associated with the key.
+   * @param overwrite_existing will overwrite the value when the value for a given key already
+   * exists.
+   * @return false when a value already exists for the given key.
+   */
+  bool add(absl::string_view key, Value value, bool overwrite_existing = true) {
+    // Check if the key already exists
+    Value existing = find(key);
+
+    // If a value exists and we shouldn't overwrite, return false
+    if (static_cast<bool>(existing) && !overwrite_existing) {
```

Comment:
Shouldn't this check be similar to what hasValue does?
(Or, does this imply that the std::is_pointer_v check in hasValue is redundant?)